### PR TITLE
Adding AngularFireAnalytics, AngularFireRemoteConfig, and refactoring DI Tokens

### PR DESCRIPTION
**Pull request by jamesdaniels**
* ~Adding support for Firebase v7~ added in 5.2.3
* Adding `AngularFireAnalytics` and `AngularFireRemoteConfig` modules
* Adding`ScreenTrackingService` and `UserTrackingService` which operate in conjunction with `AngularFireAnalytics` to track screen views and user identifiers respectively
* Cleaning up the injection token naming patterns, for semver they are aliases to the old ones

TODO:
* [x] address the Zone.js issues introduced
* [x] rethink the Remote Config API, not in love
* [x] test coverage
* [x] docs
* [x] add `startWith` support to `AngularFireRemoteConfig`
* [x] export `filterRemote()`, `filterFresh()`, etc. pipes in `@angular/fire/remote-config`
* [x] add `distinctUntilChanged` with a custom compare function to `AngularFireRemoteConfig`
* [x] export Value as `implements Partial<remoteConfig.Value>` in the .d.ts, so we don't break on minors
* [x] generally clean up the types in RC
* [x] fix the types if you are running older Firebase (this will make travis happy again)
* [x] implement more of the router types in analytics (perhaps stop using _loadedConfig...)
* [x] continue to test the robustness of analytics
* [x] expand the test coverage
* [x] fix the proxy in the SSR environment

# AngularFire "Lazy"

Both `AngularFireAnalytics` and `AngularFireRemoteConfig` are of a design @davideast and I have been thinking about for a while; that we've been calling "AngularFire Lazy".

These modules not only provide convenience observables and integrations with Angular, they also lazily load their respective Firebase modules and proxy the Firebase SDK (accounting for the aforementioned lazy-loading) and patch Zone.js.

This allows you to use all of the methods available on the Firebase SDK while knowing that AngularFire has addressed Angular-compatibility. The only major difference is that anything in the vanilla Firebase SDK that isn't a promise now is.

The use of a proxy means that we should not need to add support for new additions to the Firebase SDK for you to take advantage.

**Your feedback is wanted!** We'll be bringing similar capabilities to the other modules in future releases 😄 

# `@angular/fire/analytics`

## `AngularFireAnalyticsModule`

Provides `AngularFireAnalytics` and initializes `ScreenTrackingService` and `UserTrackingService`, if they were loaded.

## `AngularFireAnalytics`

Lazy loads `firebase/analytics` and proxies `firebase.analytics()`. `APP_VERSION` and `APP_NAME` will be loaded into the Google Analytics, if they are provided.

### API

```ts
updateConfig(options: {[key:string]: any}): Promise<void>;

// from firebase.analytics() proxy:
logEvent(eventName: string, eventParams?: {[key: string]: any}, options?: analytics.AnalyticsCallOptions): Promise<void>;
setCurrentScreen(screenName: string, options?: analytics.AnalyticsCallOptions): Promise<void>;
setUserId(id: string, options?: analytics.AnalyticsCallOptions): Promise<void>;
setUserProperties(properties: analytics.CustomParams, options?: analytics.AnalyticsCallOptions): Promise<void>;
setAnalyticsCollectionEnabled(enabled: boolean): Promise<void>;
app: Promise<app.App>;
```

### DI

`ANALYTICS_COLLECTION_ENABLED: boolean`
Globally disable Google Analytics collection by setting this to false. _(default: true)_

`APP_VERSION: string`
The application version to pass to Google Analytics.

`APP_NAME: string`
The application name to pass to Google Analytics.

`DEBUG_MODE: boolean`
Start Google Analytics in debug mode, so you can test your events in [DebugView in the Firebase Console.](https://firebase.corp.google.com/project/_/analytics/debugview) _(default: false)_

Also takes `FIREBASE_OPTIONS` and `FIREBASE_APP_NAME` like all other Modules.

### Usage

```ts
@NgModule({
  imports: [
    AngularFireModule.initializeApp(environment.firebase),
    AngularFireAnalyticsModule
  ],
  providers: [
    ScreenTrackingService,
    UserTrackingService,
    { provide: DEBUG_MODE, useFactory: isDevMode },
    { provide: APP_NAME, useValue: 'my awesome web app' },
    { provide: APP_VERSION, useValue: '0.0.0 (707e803)'}
  ]
})
export class AppModule { }

...

constructor(analytics: AngularFireAnalytics) {
    analytics.logEvent('custom_event', { ... }).then(() => {
        ...
    });
}
```

## `ScreenTrackingService`

Logs screen views and tracks the current screen on Router NavigationEnd events.

## `UserTrackingService`

Tracks the user's uid, if `firebase/auth` is loaded.

# `@angular/fire/remote-config`

## `AngularFireRemoteConfigModule`

Provides `AngularFireRemoteConfig`

## `AngularFireRemoteConfig`

Lazy loads `firebase/remote-config`, proxies `firebase.remoteConfig()`, and provides convenience observables & pipes for working with Remote Config.

### API

```ts
interface ConfigTemplate {[key:string]: string|number|boolean}

type Parameter extends remoteConfig.Value {
  key: string,
  fetchTimeMillis: number
}

class AngularFireRemoteConfig {
  changes:    Observable<Parameter>;
  parameters: Observable<Parameter[]>;
  numbers:    Observable<{[key:string]: number|undefined}>  & {[key:string]: Observable<number>};
  booleans:   Observable<{[key:string]: boolean|undefined}> & {[key:string]: Observable<boolean>};
  strings:    Observable<{[key:string]: string|undefined}>  & {[key:string]: Observable<string|undefined>};
  
  // from firebase.remoteConfig() proxy:
  activate: () => Promise<boolean>;
  ensureInitialized: () => Promise<void>;
  fetch: () => Promise<void>;
  fetchAndActivate: () => Promise<boolean>;
  getAll: () => Promise<{[key:string]: remoteConfig.Value}>;
  getBoolean: (key:string) => Promise<boolean>;
  getNumber: (key:string) => Promise<number>;
  getString: (key:string) => Promise<string>;
  getValue: (key:string) => Promise<remoteConfig.Value>;
  setLogLevel: (logLevel: remoteConfig.LogLevel) => Promise<void>;
  settings: Promise<remoteConfig.Settings>;
  defaultConfig: Promise<{[key: string]: string | number | boolean}>;
  fetchTimeMillis: Promise<number>;
  lastFetchStatus: Promise<remoteConfig.FetchStatus>;
}

// Pipes for working with .changes and .parameters

filterRemote: () => MonoTypeOperatorFunction<Parameter | Parameter[]>
filterFresh: (interval: number) => MonoTypeOperatorFunction<Parameter | Parameter[]>
budget: <T>(interval: number) => MonoTypeOperatorFunction<T>

// map and scanToObject have several overloads
// scanToObject is for use with .changes
scanToObject: () => OperatorFunction<Parameter, {[key: string]: string|undefined}>
// mapToObject is the same behavior are scanToObject but for use with .parameters,
mapToObject: () => OperatorFunction<Parameter[], {[key: string]: string|undefined}>


```

### DI

`DEFAULT_CONFIG: {[key:string]: string|number|boolean}`
Provide default values for Remote Config

`REMOTE_CONFIG_SETTINGS: remoteConfig.Settings`
Configure your remote config instance

Also takes `FIREBASE_OPTIONS` and `FIREBASE_APP_NAME` like all other Modules.

### Usage

```ts
@NgModule({
  imports: [
    AngularFireModule.initializeApp(environment.firebase),
    AngularFireRemoteConfigModule
  ],
  providers: [
    { provide: DEFAULT_CONFIG, useValue: { enableAwesome: true } },
    {
      provide: REMOTE_CONFIG_SETTINGS,
      useFactory: () => isDevMode() ? { minimumFetchIntervalMillis: 1 } : {}
    }
  ]
})
export class AppModule { }

...

constructor(remoteConfig: AngularFireRemoteConfig) {
    remoteConfig.changes.pipe(
      filterFresh(172_800_000), // ensure we have values from at least 48 hours ago
      first(),
      // scanToObject when used this way is similar to defaults
      // but most importantly smart-casts remote config values and adds type safety
      scanToObject({
        enableAwesome: true,
        titleBackgroundColor: 'blue',
        titleFontSize: 12
      })
    ).subscribe(…);

    // all remote config values cast as strings
    remoteConfig.strings.subscribe(...)
    remoteConfig.booleans.subscribe(...); // as booleans
    remoteConfig.numbers.subscribe(...); // as numbers

    // convenience for observing a single string
    remoteConfig.strings.titleBackgroundColor.subscribe(...);
    remoteConfig.booleans.enableAwesome.subscribe(...); // boolean
    remoteConfig.numbers.titleBackgroundColor.subscribe(...); // number

    // however those may emit more than once as the remote config cache fires and gets fresh values from the server
    // you can filter it out of .changes for more control:
    remoteConfig.changes.pipe(
      filter(param => param.key === 'titleBackgroundColor'),
      map(param => param.asString())
      // budget at most 800ms and return the freshest value possible in that time
      // our budget pipe is similar to timeout but won't error or abort the pending server fetch (it won't emit it, if the deadline is exceeded, but it will have been fetched so can use the freshest values on next subscription)
      budget(800),
      last()
    ).subscribe(...)

    // just like .changes, but scanned as into an array
    remoteConfig.parameters.subscribe(all => ...);

    // or make promisified firebase().remoteConfig() calls direct off AngularFireRemoteConfig
    // using our proxy
    remoteConfig.getAll().then(all => ...);
    remoteConfig.lastFetchStatus.then(status => ...);
}
```

### CODE REVIEW

**Comment by jamesdaniels** [comment]
@jhuleatt @davideast just a WIP, need to get the tests together. But feel free to start reviewing implementation and API.

**Comment by johanchouquet** [comment]
Hi @jamesdaniels ,  i find the APIs for Analytics and Remote Config quite nice. So cool to have these new features ready soon.

**Comment by NidasioAlberto** [comment]
Looking forward for this features

**Comment by Mintenker** [comment]
@jamesdaniels What's the status of this? We would really love to use it as soon as possible in our app :)

**Comment by Newbie012** [comment]
What's holding from the PR to be merged?

**Comment by jamesdaniels** [comment]
Mostly that I was taking some time off of work 😛 I'm back now, expect the tests and docs to get going here, might cut an RC for anyone interested in testing ASAP.

**Comment by jimmykane** [comment]
Interested in testing asap

**Comment by jamesdaniels** [comment]
Turns out the Zone.js issues weren't specific to this PR, I was hunting ghosts. These were fixed in [5.2.3](https://github.com/angular/angularfire/releases/tag/5.2.3). Now that these are out of the way, I only have docs to work on. Will cut an RC tomorrow.

**Comment by jamesdaniels** [comment]
Updated this branch and the comment, big changes, looking for feedback. Thanks!

**Comment by jamesdaniels** [comment]
Will cut `5.3-rc.0` shortly

**Comment by jamesdaniels** [comment]
5.3.0-rc.0 has been cut to NPM next.

File: src/analytics/analytics.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+            analytics.setCurrentScreen(screen_name || url, { global: outlet == "primary" });
+          }
+        }),
+        runOutsideAngular(zone)
+      ).subscribe();
+    }
+
+    // TODO do something other than just check auth presence, what if it's lazy loaded?
+    if (app.auth && automaticallyTrackUserIdentifier !== false) {
+      new Observable<firebase.User|null>(app.auth().onAuthStateChanged.bind(app.auth())).pipe(
```
Comment (Kamshak) [inline src/analytics/analytics.ts]: Subscription leaking here, suggest adding an ngOnDestroy to the service and unsubscribe when the service is destroyed

**Comment by Kamshak** [comment]
An Idea that is a cool alternative (IMO) to controlling the behavior through injection (e.g. AUTOMATICALLY_TRACK_USER_IDENTIFIER) is to have a service that you can inject which has only that logic in the constructor. To enable the feature you inject it in your AppComponent, if you don't want it you don't inject it, so it never gets created and the unused feature can be tree shaken out.

Angulartics2 does it this way

File: src/analytics/analytics.ts (hunk 1 above)
Comment (jamesdaniels) [reply src/analytics/analytics.ts]: Totally. Thanks.

**Comment by jamesdaniels** [comment]
@Kamshak I was debating a service vs. built into the module & when/where I should start drawing the line as I increase built-in utilities. It being tree shakable is a good point in favor of services. Appreciate the feedback!

File: src/remote-config/tsconfig-test.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
+{
+  "extends": "./tsconfig-esm.json",
+  "compilerOptions": {
+    "baseUrl": ".",
+    "paths": {
+      "@angular/fire": ["../../dist/packages-dist"],
+      "@angular/fire/reomte-config": ["../../dist/packages-dist/remote-config"]
```
Comment (andipaetzold) [inline src/remote-config/tsconfig-test.json]: Typo

```suggestion
      "@angular/fire/remote-config": ["../../dist/packages-dist/remote-config"]
```

**Comment by jimmykane** [comment]
Why is the default app name `Angular App`?

**Comment by jamesdaniels** [comment]
> Why is the default app name Angular App?

🤷‍♂ any suggestions?

File: src/remote-config/tsconfig-test.json (hunk 2 above)
Comment (jamesdaniels) [reply src/remote-config/tsconfig-test.json]: Thanks

**Comment by jamesdaniels** [comment]
@andipaetzold could you look into the Googlebot's complaint? Want to make sure that way we're dotting our Is and crossing our Ts.

**Comment by andipaetzold** [comment]
@googlebot I fixed it.

**Comment by jamesdaniels** [comment]
5.3.0-rc.1 has been cut to NPM next. See the updated PR description.

Big changes are the introduction of the `UserTrackingService` and `ScreenTrackingService`, tweaked RC API, and retooling of the screen tracking to give you a more robust analytics experience in the Firebase Console & BigQuery.

File: src/analytics/analytics.service.ts [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+            tap(params => {
+                // TODO perhaps I can be smarter about this, bubble events up to the nearest outlet?
+                if (params.outlet == "primary") {
+                    // TODO do I need to add gtag config for firebase_screen, firebase_screen_class, firebase_screen_id?
+                    //      also shouldn't these be computed in the setCurrentScreen function? prior too?
+                    //      do we want to be logging screen name or class?
+                    analytics.setCurrentScreen(params.screen_name, { global: true })
+                }
+            }),
+            map(params => ({ firebase_screen_id: nextScreenId(params), ...params})),
```
Comment (jamesdaniels) [inline src/analytics/analytics.service.ts]: @Feiyang1 @hsubox76 what would be the interest level in my moving some the `firebase_*` parameter handling (`firebase_screen_id`, `firebase_previous_id`, `firebase_screen_class`, etc.) into the JS SDK?

I know it breaks the separation of concerns a bit but was non-trivial to get working correctly. It's needed to be able to track user flows in the Firebase Console and BigQuery. The iOS and Android SDKs do this automatically for the developer and their behavior is what I modeled here for Angular. I'll be doing the same treatment for React and Ember routers. I suspect other 3p SDKs will been keen on implementing too.

At the very least we should add information on how to implement in the Firebase docs and add types. FYI @gkaldev @schandel @davideast

Comment (hsubox76) [reply src/analytics/analytics.service.ts]: I don't have a lot of Angular experience so maybe wait for @feiyang1 to take a look but a lot of this looks very Angular-router-specific.  What part/params would be passed to the JS SDK to work with? Do we derive screen_class and previous_id from something or would that be passed in?

Comment (jamesdaniels) [reply src/analytics/analytics.service.ts]: Hooking into the router will be framework specific of course, my thought would be to add a `logScreenView(this, name: string, additionalParams?: {})` function which would do all the firebase specific event param computation: `firebase_screen_id`, `firebase_previous_screen`, `firebase_screen_class`, `firebase_screen`, `firebase_previous_class`, `firebase_previous_screen`, and `firebase_previous_id`. They're are all undocumented and having to do these on my own to get the most out of Firebase was unexpected. 

For instance `firebase_screen_id` challenging to understand, I couldn't find any resources externally and had to "cheat" and look it up on g3 to find out it's a random INT64 (rather than a hashCode of some kind). I'm not 100% confident I got it right either, hence looping in more DelRel folk.

Comment (hsubox76) [reply src/analytics/analytics.service.ts]: Yeah, if these are standard for other Firebase platforms and we can come up with a sort of general interface I think it would make sense for us to bake some of this handling into the JS SDK.  Let us know if you want to set up a meeting to talk about what we can do with these and pass along your hard-earned research about these params, maybe after you get some more input from other devrels confirming how they work.

Comment (Feiyang1) [reply src/analytics/analytics.service.ts]: IIUC, you are asking the JS SDK to keep the state of the previous `screen_view` parameters and attach them to the next `screen_view` request?
Can you share a pointer to how iOS SDK and Android SDK handles it?

**Comment by doczoidberg** [comment]
I installed 5.3.0-rc.1 and added AngularAnalytics as described here: https://github.com/angular/angularfire/blob/62d90b919100f4086c27ce05230e42a47a3c6c05/docs/analytics/getting-started.md

But also after 24 hours no data is shown in my firebase console.

File: src/core/angularfire2.ts [hunk 4]
```diff
@@ -65,3 +67,36 @@ export const runInZone = (zone: NgZone) => <T>(obs$: Observable<T>): Observable<
...
+        // TODO move to proper type guards
+        if (typeof ret == 'function') {
+          return ret.bind(mod);
+        } else if (ret && ret.then) {
+          return ret.then((res:any) => zone.run(() => res));
+        } else {
+          return zone.run(() => ret);
+        }
+      }), {
+        get: (self, name) => self()[name],
```
Comment (Feiyang1) [inline src/core/angularfire2.ts]: Please bear with my ignorance, shouldn't it be `get: (self, name) => self().then(it => it[name])`, since `self()` returns a promise?

Comment (jamesdaniels) [reply src/core/angularfire2.ts]: Probably unneeded but the get here is a passthrough onto the attributes of the promise, say you wanted to get `.constructor` or `.then` without calling the function. My suspicion here is that this will be useful for type-checking or async/await.

Triggering `apply` when a function is called is why I went higher order on the first arg to the proxy.

However, I'm planning on recursing the get (and apply) here so: `self()[name] || _lazySDKProxy(klass, from(self()), zone)` which would allow chaining: `await angularFirestore.collection('asdf').ref` rather than `await angularFirestore.collection('asdf').then(c => c.ref)`

**Comment by jamesdaniels** [comment]
> I installed 5.3.0-rc.1 and added AngularAnalytics as described here: https://github.com/angular/angularfire/blob/62d90b919100f4086c27ce05230e42a47a3c6c05/docs/analytics/getting-started.md
> But also after 24 hours no data is shown in my firebase console.

Interesting, thanks for the feedback. I'm also not seeing the analytics I'd like from this. I'll continue to work on this.

File: src/analytics/analytics.service.ts (hunk 3 above)
Comment (jamesdaniels) [reply src/analytics/analytics.service.ts]: Yeah, basically I'm thinking all the undocumented `firebase_*` attributes that make the console work should be something we take care of.

I'll send over email now that I'm back from the holiday and we can hop on GVC if additional clarity is needed. That probably would've been a better place for the discussion anyway. 😛

Comment (jamesdaniels) [reply src/analytics/analytics.service.ts]: My Firebase console is still not showing the information I expect, meaning I did not get this 100% correct yet.

**Comment by doczoidberg** [comment]
@jamesdaniels PR was 2 month ago. How long do we have to wait to get this in production?

**Comment by leolara** [comment]
Hi there
one note, it seems from here https://developers.google.com/analytics/devguides/collection/app-web/screen-view the only parameter expected for `screen_view` is `screen_name`

If that is so, perhaps, the code can be reduced?

**Comment by jamesdaniels** [comment]
> Hi there
> one note, it seems from here https://developers.google.com/analytics/devguides/collection/app-web/screen-view the only parameter expected for `screen_view` is `screen_name`
> 
> If that is so, perhaps, the code can be reduced?

For GA legacy yes. The Firebase experience is richer, which [you can see on the Firebase Demo Project](https://firebase.corp.google.com/project/fir-demo-project/analytics/app/android:com.labpixies.flood/events~2Freport~2Fscreen_view%3Ft=1575925132106&fpn=300830567303&swu=1&sgu=1&sus=upgraded&params=_u..pageSize%253D25%2526_r..eventId%253Dscreen_view&cs=app.m.events.detail&r=screen_view&g=1). In order to get all the great screen tracking metrics the Firebase Android and iOS SDKs track a bunch more details for you, note all the `firebase_` attributes. I'm trying to give the same benefits to Angular users here.

**Comment by jamesdaniels** [comment]
rc.2 is up on NPM. There’s two or three small things I need to address beyond docs but I’m happy with the API. Note, while the analytics do appear to be going up correctly there seem to be bugs in the Firebase Console that are keeping version uptake and class names from appearing where expected. I’m filing issues there.

**Comment by doczoidberg** [comment]
Installed RC.2 but get still no analytics data in my firebase console

**Comment by jimmykane** [comment]
For the info , looks like firebase analytics has issues. I don't get
realtime visits but I do get all the daily etc.

This has been broken recently without updating this lib.

@James maybe something for the Firebase team ?

**Comment by jamesdaniels** [comment]
@jimmykane I’ve noticed this since about Wednesday/Thursday too. Debug view doesn’t seem to respond either... I’ll investigate.

**Comment by LanderBeeuwsaert** [comment]
yeah, has nothing to do with angularFire, I use analytics with the normal firebase library and I have it as well.

**Comment by jimmykane** [comment]
@jamesdaniels  yeah around that time.  

For me the debug view never worked but I just use the AFA "normal" calls. Not sure if it should work then...

**Comment by jimmykane** [comment]
hmmm I opened a Firebase support ticket and got back: 

```
Currently, Google Analytics for Firebase for Web only supports JavaScript. As this library is not created nor maintained by the Firebase engineers. 
```

**Comment by jamesdaniels** [comment]
@jimmykane just for clarification (in case you weren't aware) we are Firebase engineers and support this library on a best-effort basis. Can you forward the relevant case details to my email, jamesdaniels@google.com, so I can look into how that case was handled?

**Comment by jamesdaniels** [comment]
New RC API has been published, `5.3.0-rc.3` on NPM under the `@next` tag. Unless something major comes up, going to call this the final RC for this PR and release later this week.

Rocking out on docs is the plan from here and making sure the proxy doesn't explode when you try to make a call in a non-browser environment.

Thanks for all the feedback!

Gotta ship this and clear my plate so I can immediately jump on Angular 9 and cut AngularFire v6 after the holidays!

**Comment by savilaf** [comment]
We are getting this error with the `5.3.0-rc.3` version:
```
ErrorHandler: Error: Uncaught (in promise): FirebaseError: Remote Config: The config fetch request timed out while in an exponential backoff state. Configure timeout using "fetchTimeoutMillis" SDK setting. Unix timestamp in milliseconds when fetch request throttling ends: 1576830408459. (remoteconfig/fetch-throttle).
FirebaseError: Remote Config: The config fetch request timed out while in an exponential backoff state. Configure timeout using "fetchTimeoutMillis" SDK setting. Unix timestamp in milliseconds when fetch request throttling ends: 1576830408459. (remoteconfig/fetch-throttle).
    at index.cjs.js:1067
    at index.cjs.js:373
    at Array.forEach (<anonymous>)
    at RemoteConfigAbortSignal.push../node_modules/@firebase/remote-config/dist/index.cjs.js.RemoteConfigAbortSignal.abort (index.cjs.js:373)
    at RemoteConfig.<anonymous> (index.cjs.js:558)
    at step (tslib.es6.js:99)
    at Object.next (tslib.es6.js:80)
    at tslib.es6.js:73
    at new ZoneAwarePromise (zone-evergreen.js:876)
    at Module.__awaiter (tslib.es6.js:69)
    at resolvePromise (zone-evergreen.js:797)
    at zone-evergreen.js:707
    at zone-evergreen.js:723
    at ZoneDelegate.invoke (zone-evergreen.js:359)
    at Object.onInvoke (core.js:39699)
    at ZoneDelegate.invoke (zone-evergreen.js:358)
    at Zone.run (zone-evergreen.js:124)
    at zone-evergreen.js:855
    at ZoneDelegate.invokeTask (zone-evergreen.js:391)
    at Object.onInvokeTask (core.js:39680)
```
I have not been able to understand what is failing exactly but RemoteConfig is working as expected.
The error appears after a period of inactivity.

**Comment by jimmykane** [comment]
I think I have an issue or this is due to AF RC + Angular 9
I get: 

```
Access to fetch at 'https://firebaseinstallations.googleapis.com/v1/projects/quantified-self-io/installations/cfw8xt-wFML3zmZ-Yp-hbi/authTokens:generate' from origin 'https://beta.quantified-self.io' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.```

**Comment by jamesdaniels** [comment]
Cutting `5.3.0-rc.4`. Turns out `screen_class` wasn't tracking to my satisfaction in lazy loaded routes. I'm out of the well trodden path of public APIs (that I know about) and I haven't tried to test this with deeply nested / complex routes or pushed on trickier edge cases in the Angular router, so would appreciate eyeballs on the `ScreenTrackingService`.

**Comment by jamesdaniels** [comment]
@savilaf sounds like the RC endpoint is just throttling you, this exception should affect anything other than your possibly getting more fresh values from the server. You can fix this error by either increasing your `minimumFetchIntervalMillis` in `REMOTE_CONFIG_SETTINGS` or using the `filterFresh` pipe I provide in addition to a `first`.

See how I addressed in one of my apps here:

```ts
export class AppComponent implements OnInit, OnDestroy {

  public configuration$: Observable<typeof environment.remoteConfigDefaults & Record<string, string|undefined>>;

  constructor(
    private remoteConfig: AngularFireRemoteConfig,
    @Inject(PLATFORM_ID) private platformId: Object
  ) { }

  ngOnInit() {
    const template = environment.remoteConfigDefaults;
    this.configuration$ = this.remoteConfig.parameters.pipe(
      untilComponentDestroyed(this),
      // rate-limit to 48-hours, unless we're not in the browser & only have defaults
      isPlatformBrowser(this.platformId) ? filterFresh(172_800_000) : tap(),
      first(),
      mapToObject(template)
    );
  }

  ngOnDestroy() { }

}
```

**Comment by mattgek** [comment]
Hi,

Will there be an option to disable analytics (GDPR Europe)? It must be possible for users to opt-out from analytics.

**Comment by jamesdaniels** [comment]
@mattgek tottally. With this API you can make it opt-in via setting `ANALYTICS_COLLECTION_ENABLED` to false by default, then opt-in with `setAnalyticsCollectionEnabled(true)`.

Opt-out you can simply `setAnalyticsCollectionEnabled(false)`.

Either way you can then `useFactory` for `ANALYTICS_COLLECTION_ENABLED` DI token and look at the user's cookies or local store; so their choice persists.

**Comment by jimmykane** [comment]
On Angular 9 I get when I enable the `UserTrackingService` :

```
ERROR TypeError: Cannot read property 'subscribe' of undefined
    at Observable._subscribe (subscribeToObservable.js:4)
    at Observable._trySubscribe (Observable.js:42)
    at Observable.subscribe (Observable.js:28)
    at MapOperator.call (map.js:16)
    at Observable.subscribe (Observable.js:23)
    at SwitchMapOperator.call (switchMap.js:17)
    at Observable.subscribe (Observable.js:23)
    at SwitchMapOperator.call (switchMap.js:17)
    at Observable.subscribe (Observable.js:23)
    at Observable._subscribe (angularfire2.js:59)
    at Observable._trySubscribe (Observable.js:42)
    at Observable.subscribe (Observable.js:28)
    at angularfire2.js:53
    at ZoneDelegate.invoke (zone-evergreen.js:365)
    at Zone.run (zone-evergreen.js:124)
    at NgZone.runOutsideAngular (core.js:41316)
    at Observable._subscribe (angularfire2.js:52)
    at Observable._trySubscribe (Observable.js:42)
    at Observable.subscribe (Observable.js:28)
    at new UserTrackingService (analytics.service.js:138)
    at Object.UserTrackingService_Factory [as factory] (analytics.service.js:146)
    at R3Injector.hydrate (core.js:18772)
    at R3Injector.get (core.js:18534)
    at injectInjectorOnly (core.js:903)
    at Module.ɵɵinject (core.js:913)
    at Object.AngularFireAnalyticsModule_Factory [as factory] (analytics.module.js:21)
    at R3Injector.hydrate (core.js:18772)
    at R3Injector.get (core.js:18534)
    at core.js:18464
    at Set.forEach (<anonymous>)
    at new R3Injector (core.js:18460)
    at createInjector (core.js:18411)
    at new NgModuleRef$1 (core.js:36393)
    at NgModuleFactory$1.create (core.js:36496)
    at core.js:42390
    at ZoneDelegate.invoke (zone-evergreen.js:365)
    at Object.onInvoke (core.js:41480)
    at ZoneDelegate.invoke (zone-evergreen.js:364)
    at Zone.run (zone-evergreen.js:124)
    at NgZone.run (core.js:41255)
    at PlatformRef.bootstrapModuleFactory (core.js:42383)
    at core.js:42459
    at ZoneDelegate.invoke (zone-evergreen.js:365)
    at Zone.run (zone-evergreen.js:124)
    at zone-evergreen.js:851
    at ZoneDelegate.invokeTask (zone-evergreen.js:400)
    at Zone.runTask (zone-evergreen.js:168)
    at drainMic
```

File: src/firestore/firestore.ts [hunk 5]
```diff
@@ -17,13 +17,19 @@ import firebase from 'firebase/app';
...
 /**
  * The value of this token determines whether or not the firestore will have persistance enabled
  */
 export const EnablePersistenceToken = new InjectionToken<boolean>('angularfire2.enableFirestorePersistence');
 export const PersistenceSettingsToken = new InjectionToken<PersistenceSettings|undefined>('angularfire2.firestore.persistenceSettings');
 export const FirestoreSettingsToken = new InjectionToken<Settings>('angularfire2.firestore.settings');
 
+export const ENABLE_PERSISTENCE = EnablePersistenceToken;
+export const PERSISTENCE_SETTINGS = PersistenceSettingsToken
+export const FIRESTORE_SETTINGS = FirestoreSettingsToken;
```
Comment (jamesdaniels) [inline src/firestore/firestore.ts]: Change to just `SETTINGS`

File: src/storage/storage.ts [hunk 6]
```diff
@@ -2,11 +2,13 @@ import { Injectable, Inject, Optional, InjectionToken, NgZone, PLATFORM_ID } fro
...
 import { createUploadTask, AngularFireUploadTask } from './task';
 import { Observable } from 'rxjs';
-import { FirebaseStorage, FirebaseOptions, FirebaseAppConfig, FirebaseOptionsToken, FirebaseNameOrConfigToken, FirebaseZoneScheduler, _firebaseAppFactory } from '@angular/fire';
+import { FirebaseStorage, FirebaseOptions, FirebaseAppConfig, FirebaseZoneScheduler, _firebaseAppFactory, FIREBASE_OPTIONS, FIREBASE_APP_NAME } from '@angular/fire';
 
 import { UploadMetadata } from './interfaces';
 
+// SEMVER drop StorageBucket in favor of STORAGE_BUCKET
 export const StorageBucket = new InjectionToken<string>('angularfire2.storageBucket');
+export const STORAGE_BUCKET = StorageBucket;
```
Comment (jamesdaniels) [inline src/storage/storage.ts]: Just change to `BUCKET`

File: src/core/angularfire2.ts [hunk 7]
```diff
@@ -3,7 +3,9 @@ import { isPlatformServer } from '@angular/common';
 import { Observable, Subscription, queueScheduler as queue } from 'rxjs';
 
 // Put in database.ts when we drop database-depreciated
+// SEMVER drop RealtimeDatabaseURL in favor of DATABASE_URL in next major
 export const RealtimeDatabaseURL = new InjectionToken<string>('angularfire2.realtimeDatabaseURL');
+export const DATABASE_URL = RealtimeDatabaseURL;
```
Comment (jamesdaniels) [inline src/core/angularfire2.ts]: Reexport as simply `URL` in `database` and `database-depreciated`

**Comment by jamesdaniels** [comment]
Released 5.3 with some minor changes from the last RC. Thanks for your feedback everyone, if you’re still having issues please file new issues. 

The priority from here will be getting ng9 / ivy support all buttoned up; and the backlog of PRs addresses.
